cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -676,44 +687,55 @@ public void close() throws IOException {
 
   @Override
   public String getInputFormatClassName() {
-    Collection<String> inputPaths = filterInputPaths();
-    if (inputPaths != null && inputPaths.isEmpty()) {
+    Collection<PartitionKey> inputKeys = getInputKeys();
+    if (inputKeys.isEmpty()) {
       return EmptyInputFormat.class.getName();
     }
     return files.getInputFormatClassName();
   }
 
   @Override
   public Map<String, String> getInputFormatConfiguration() {
-    Collection<String> inputPaths = filterInputPaths();
-    if (inputPaths == null) {
-      return files.getInputFormatConfiguration();
+    Collection<PartitionKey> inputKeys = getInputKeys();
+
+    List<Location> inputLocations = new ArrayList<>(inputKeys.size());
+    Map<String, PartitionKey> pathToKey = new HashMap<>(inputKeys.size());
+    for (PartitionKey key : inputKeys) {
+      PartitionDetail partition = getPartition(key);
+      String path = Objects.requireNonNull(partition).getRelativePath();
+      Location partitionLocation = files.getLocation(path);
+      inputLocations.add(partitionLocation);
+      pathToKey.put(partitionLocation.toURI().toString(), key);
     }
-    List<Location> inputLocations = Lists.newArrayListWithExpectedSize(inputPaths.size());
-    for (String path : inputPaths) {
-      inputLocations.add(files.getLocation(path));
-    }
-    return files.getInputFormatConfiguration(inputLocations);
+
+    Map<String, String> inputFormatConfiguration = files.getInputFormatConfiguration(inputLocations);
+    inputFormatConfiguration.put(PATH_TO_PARTITIONING_MAPPING, GSON.toJson(pathToKey));
```

Comment:
It uses java's Properties class internally, and I didn't find any limit on that.

I tested with 10 million+ characters in the value of one key, and the MapReduce job (PurchaseHistoryBuilder) still worked.
As expected, the xml file grew, but other than that, I didn't notice any issue:
```
ls -ltrh /data/yarn/local/usercache/cdap/appcache/application_1479510137122_0055/filecache/13/job.xml
-r-x------. 1 cdap cdap 9.9M Nov 24 00:27 /data/yarn/local/usercache/cdap/appcache/application_1479510137122_0055/filecache/13/job.xml
```

Here's how I modified PurchaseHistoryBuilder:
```java
--- a/cdap-examples/Purchase/src/main/java/co/cask/cdap/examples/purchase/PurchaseHistoryBuilder.java
+++ b/cdap-examples/Purchase/src/main/java/co/cask/cdap/examples/purchase/PurchaseHistoryBuilder.java
@@ -75,6 +75,15 @@ public class PurchaseHistoryBuilder extends AbstractMapReduce {
     if (reducerMemoryMBStr != null) {
       context.setReducerResources(new Resources(Integer.parseInt(reducerMemoryMBStr)));
     }
+
+    String tenChars = "0123456789";
+    StringBuilder stringBuilder = new StringBuilder();
+    String conflen = context.getRuntimeArguments().get("conflen");
+    for (int i = 0; i < Integer.parseInt(conflen); i++) {
+      stringBuilder.append(tenChars);
+    }
+    stringBuilder.append(tenChars);
+    job.getConfiguration().set("long.key", stringBuilder.toString());
   }
 
```